controllers/noderefutil/indexer.go

```diff
@@ -65,14 +66,46 @@ func IndexNodeByProviderID(o client.Object) []string {
 		panic(fmt.Sprintf("Expected a Node but got a %T", o))
 	}
 
-	if node.Spec.ProviderID != "" {
-		providerID, err := NewProviderID(node.Spec.ProviderID)
-		if err != nil {
-			// Failed to create providerID, skipping.
-			return nil
-		}
-		return []string{providerID.IndexKey()}
+	if node.Spec.ProviderID == "" {
+		return nil
+	}
+
+	providerID, err := NewProviderID(node.Spec.ProviderID)
+	if err != nil {
+		// Failed to create providerID, skipping.
+		return nil
+	}
+	return []string{providerID.IndexKey()}
+}
+
+// AddMachineProviderIDIndex adds the machine providerID index to the
+// managers cache.
+func AddMachineProviderIDIndex(ctx context.Context, mgr ctrl.Manager) error {
+	if err := mgr.GetCache().IndexField(ctx, &clusterv1.Machine{},
+		clusterv1.MachineProviderIDIndex,
+		IndexMachineByProviderID,
+	); err != nil {
+		return errors.Wrap(err, "error setting index fields")
```

Comment:
```suggestion
		return errors.Wrap(err, "error setting index field")
```
nit
(same in the original above)